render-ws-java-client/src/main/java/org/janelia/render/client/MultiSemFlatFieldCorrectionClient.java

```diff
@@ -0,0 +1,195 @@
+package org.janelia.render.client;
+
+import com.beust.jcommander.Parameter;
+import com.beust.jcommander.ParametersDelegate;
+import ij.IJ;
+import ij.ImagePlus;
+import ij.process.FloatProcessor;
+import ij.process.ImageProcessor;
+import org.janelia.alignment.ImageAndMask;
+import org.janelia.alignment.loader.ImageLoader;
+import org.janelia.alignment.spec.ChannelSpec;
+import org.janelia.alignment.spec.ResolvedTileSpecCollection;
+import org.janelia.alignment.spec.TileSpec;
+import org.janelia.alignment.spec.stack.StackId;
+import org.janelia.alignment.spec.stack.StackMetaData;
+import org.janelia.alignment.util.ImageProcessorCache;
+import org.janelia.render.client.parameter.CommandLineParameters;
+import org.janelia.render.client.parameter.MultiProjectParameters;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.DoubleSummaryStatistics;
+import java.util.List;
+
+/**
+ * Apply a previously computed flat field estimate to correct shading in a Multi-SEM project.
+ * Originally developed for the wafer_53 Multi-SEM dataset.
+ *
+ * @author Michael Innerberger
+ */
+public class MultiSemFlatFieldCorrectionClient {
+
+	// make cache large enough to hold all flat field estimates for one layer
+	private static final ImageProcessorCache CACHE = new ImageProcessorCache(91L * 2000L * 1748L + 1, false, false);
+
+	private final Parameters params;
+	private final RenderDataClient renderClient;
+
+	public static class Parameters extends CommandLineParameters {
+		@ParametersDelegate
+		private final MultiProjectParameters multiProject = new MultiProjectParameters();
+		@Parameter(names = "--flatFieldLocation", description = "Location of the flat field estimates", required = true)
+		private String flatFieldLocation;
+		@Parameter(names = "--outputRoot", description = "Folder to write corrected images to", required = true)
+		private String outputFolder;
+		@Parameter(names = "--flatFieldFormat", description = "Format of the flat field estimates in printf style with placeholders for z-layer and sfov", required = true)
+		private String flatFieldFormat;
+		@Parameter(names = "--targetStackSuffix", description = "Suffix to append to the stack name for the corrected stack", required = true)
+		private String targetStackSuffix = "_corrected";
+		@Parameter(names = "--inputRoot", description = "Root folder for input data; if given, the structure under this root is replicated in the output folder")
+		private String inputRoot = null;
+		@Parameter(names = "--flatFieldConstantFromZ", description = "Maximum z-layer of flat field estimates to consider. All subsequent z-layers get corrected with the maxium, since the estimates can be bad for the last few z-layers. If not given, all z-layers are considered")
```

Comment:
Should this be something like `--finalZLayersToIgnore` instead with a default value of zero instead since the multi-SEM stacks have different total z layer counts?

Reply:
The idea is that the flat field estimates for the last couple of z-layers are bad, because there's only very little data. Thus, Preibisch suggested to just take the last z-layer which has a decent estimate for all subsequent layers (i.e., keep the flat field constant from a particular z-layer, hence the name).
I struggled to find a good name for the parameter and a concise description. If you have a better suggestion, I'm very open to that.

Reply:
Makes sense - my guess was that you don't always want say z 35, but instead want maxZ - 4 because maxZ can range from 3? to 4?